lib/letsencrypt.js

```diff
@@ -1,215 +1,104 @@
-                                    */
-                                }
-                            },
-                            err => {
-                                err && log.error(err);
-                                cb(null, defaultTls);
-                            });
-                        } else {
-                            cb(null, hostTls);
-                        }
+                // prepare domains
+                if (typeof leSettings.domains === 'string') {
+                    leSettings.domains = leSettings.domains.split(',');
+                    for (let d = leSettings.domains.length - 1; d >= 0; d--) {
+                        leSettings.domains[d] = leSettings.domains[d].trim();
+                        // delete empty
+                        !leSettings.domains[d] && leSettings.domains.splice(d, 1);
                     }
-                } else {
-                    cb(null, defaultTls);
                 }
-            };
+                // TODO? Tried adding sites in the gl create but that didn't work so is below with gl.add.
+
+                const pkg = require('../package.json');
+                const gl = require('greenlock').create({
+                    packageRoot: __dirname,
+                    configDir: leDir,
+                    maintainerEmail: leSettings.email,
+                    cluster: false,
+                    packageAgent: pkg.name + '/' + pkg.version,
+                    notify: function (event, details) {
+                        // TODO: separate out errors, etc.
+                        log.debug(`Greenlock notify ${event}: ${details}`);
+                    }
+                });
+                gl.manager.defaults({
+                    agreeToTerms: true,
+                    subscriberEmail: leSettings.email
+                }).then(() => {
+                    // Add sites.
+                    // TODO: something seems fishy with the timing of this because on the first run
+                    // a warning about no sites being configured is found, even though things go on to
+                    // work perfectly.
+                    for (const domain of leSettings.domains) {
+                        log.debug(`Adding to Greenlock: ${domain}`);
+                        gl.add({ subject: domain, altnames: [domain] });
+                    }
 
-            try {
-                server = require('https').createServer(options, lex.middleware(app));
-            } catch (err) {
-                log.error('HTTPS server could not be started: ' + err);
-            }
-        } else {
-            try {
-                server = require('https').createServer(certificates, app);
-            } catch (err) {
-                log.error('HTTPS server could not be started: ' + err);
+                    require('greenlock-express')
+                        .init({
+                            greenlock: gl
+                        }).ready(glx => {
```

Comment:
Is there ever a case where we get something else than `ready()`, e.g. an error? If so, we should handle this.

Reply:
The Greenlock `init` function is pretty simple so unlikely. Technically it could return an object that didn't have a `ready`. In that case trying to call it will throw an error that will be caught by the caller.

I guess we could wrap this block in a `try... catch` but I don't see the point as the caller should be catching any errors already.

Reply:
A `try ... catch` would be good. I just want to prevent there being a case where the promsie doesn't resolve (or rather: error).

Reply:
Sorry, but I really don't see the point. You note this isn't a Promise, but just a regular function that returns an object that should have a `ready` member that happens to be a function. Like I said, if that member is missing or not a function any error will be caught by the caller.

Reply:
Isn't the callback called asynchronously? If so, the exception will get lost or worse: it will be an uncaught error that will crash the process. And in any case: the "top" Promise will never "end" (see above).

Reply:
This isn't a callback. Have a look at the Greenlock code.

Reply:
OK, right, agreed.